checklist.py

```diff
@@ -144,6 +146,71 @@ def _system_call(self, cmd):
         return stdout, stderr, return_value
 
 
+class CopyrightHeadersValidator(RepoValidator):
+    """Flag library files with non-standard copyright headers
+
+    See the current standard for scikit-bio's copyright headers at
+    ``http://scikit-bio.org/docs/latest/development/new_module.html``
+
+    Parameters
+    ----------
+    skip_dirs : iterable of str, optional
+        Directory names to skip during validation. Defaults to skipping any
+        directories named ``'data'`` or ``'__pycache__'`` (and anything
+        contained within them).
+
+    """
+
+    reason = "Files non-conforming to standard copyright headers"
+    COPYRIGHT_HEADER = """\
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+"""
+
+    def __init__(self, skip_dirs=None):
+        if skip_dirs is None:
+            skip_dirs = {'data', '__pycache__'}
+        self.skip_dirs = set(skip_dirs)
+
+    def _validate(self, root, dirs, files):
+        for skip_dir in self.skip_dirs:
+            if skip_dir in dirs:
+                dirs.remove(skip_dir)
+
+        invalid_files = []
+        for _file in files:
+            if not _file.endswith('.py'):
+                continue
+            pos = 0
+            f = open(os.path.join(root, _file))
+            tokens = list(tokenize.generate_tokens(f.readline))
+
+            # A shebang line or module docstring is described using just two
+            # tokens: the main string and a terminating newline. By convention,
+            # however, these lines are always followed by a newline and hence
+            # we must advance by 3 positions to get to next logical line.
+            if tokens[pos][1].startswith('#!'):
```

Comment:
That also works :) They are only advisable for scripts being executed on
the command line, and only necessary if executing without an explicit call
to the interpreter, IIRC

On Thu, Mar 5, 2015 at 9:10 AM, Jai Ram Rideout <notifications@github.com>
wrote:

> In checklist.py
> <https://github.com/biocore/scikit-bio/pull/862#discussion_r25875420>:
>
> > +            if skip_dir in dirs:
> > +                dirs.remove(skip_dir)
> > +
> > +        invalid_files = []
> > +        for _file in files:
> > +            if not _file.endswith('.py'):
> > +                continue
> > +            pos = 0
> > +            f = open(os.path.join(root, _file))
> > +            tokens = list(tokenize.generate_tokens(f.readline))
> > +
> > +            # A shebang line or module docstring is described using just two
> > +            # tokens: the main string and a terminating newline. By convention,
> > +            # however, these lines are always followed by a newline and hence
> > +            # we must advance by 3 positions to get to next logical line.
> > +            if tokens[pos][1].startswith('#!'):
>
> Can we explicitly disallow shebang lines in scikit-bio library code? We
> currently don't have any under the skbio/ directory (only checklist.py
> and setup.py have them).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/862/files#r25875420>.
>